python-frontend/src/main/java/org/sonar/python/tree/PythonTreeMaker.java

```diff
@@ -1676,62 +1667,93 @@ private static Expression numericLiteral(AstNode astNode) {
     return new NumericLiteralImpl(toPyToken(astNode.getToken()));
   }
 
-  private Expression stringLiteral(AstNode astNode) {
-    List<StringElement> elements = new ArrayList<>();
-    for (AstNode elementNode : astNode.getChildren(PythonTokenType.STRING)) {
-      com.sonar.sslr.api.Token token = elementNode.getToken();
-      StringElementImpl element = new StringElementImpl(toPyToken(token));
-      if (element.isInterpolated()) {
-        F_STRING_PARSER.fStringExpressions(token).forEach(
-          expressionNode -> element.addFormattedExpression(formattedExpression(expressionNode))
-        );
-      }
-      elements.add(element);
-    }
+  private Expression stringLiterals(AstNode astNode) {
+    List<StringElement> elements = astNode.getChildren(PythonTokenType.STRING, PythonGrammar.FSTRING).stream()
+      .map(this::stringLiteral)
+      .filter(Objects::nonNull)
+      .collect(Collectors.toList());
     return new StringLiteralImpl(elements);
   }
 
+  private StringElementImpl stringLiteral(AstNode elementNode) {
+    Token token = toPyToken(elementNode.getToken());
+    if (token == null) {
+      return null;
+    }
+    if (elementNode.is(PythonGrammar.FSTRING)) {
+      Token fstringEnd = toPyToken(elementNode.getFirstChild(PythonTokenType.FSTRING_END).getToken());
+      List<Tree> fStringMiddles = getFStringMiddles(elementNode);
+      return new StringElementImpl(token, fStringMiddles, fstringEnd);
+    }
+    return new StringElementImpl(token, List.of(), null);
+  }
+
+  private List<Tree> getFStringMiddles(AstNode expressionNode) {
+    return expressionNode
+      .getChildren(PythonGrammar.FSTRING_REPLACEMENT_FIELD, PythonTokenType.FSTRING_MIDDLE)
+      .stream()
+      .map(fStringMiddle -> {
+        if (fStringMiddle.is(PythonGrammar.FSTRING_REPLACEMENT_FIELD)) {
+          return formattedExpression(fStringMiddle);
+        }
+        return stringLiteral(fStringMiddle);
+      })
+      .filter(Objects::nonNull)
+      .collect(Collectors.toList());
+  }
+
   private FormatSpecifier formatSpecifier(AstNode expressionNode) {
     AstNode formatSpecifierNode = expressionNode.getFirstChild(PythonGrammar.FORMAT_SPECIFIER);
     if (formatSpecifierNode == null) {
       return null;
     }
 
     Token columnToken = toPyToken(formatSpecifierNode.getFirstChild(PythonPunctuator.COLON).getToken());
-    List<FormattedExpression> nestedExpressions = formatSpecifierNode.getChildren(PythonGrammar.FORMATTED_EXPR)
-      .stream()
-      .map(this::formattedExpression)
-      .collect(Collectors.toList());
-
-    return new FormatSpecifierImpl(columnToken, nestedExpressions);
+    if (columnToken == null) {
+      return null;
+    }
+    List<Tree> fStringMiddles = getFStringMiddles(formatSpecifierNode);
```

Comment:
Minor: move this assignment before column token assignment (just to group variables)